state/migration_import.go

```diff
@@ -1345,20 +1388,99 @@ func deduceCharmOrigin(url *charm.URL) (*CharmOrigin, error) {
 		t = "charm"
 	}
 
-	origin := &CharmOrigin{
+	source, channel := getApplicationSourceChannel(a, url)
+	return &CharmOrigin{
 		Type:     t,
+		Source:   source.String(),
 		Revision: &url.Revision,
-	}
+		Channel:  channel,
+		Platform: i.getApplicationPlatform(a, url, units),
+	}, nil
+}
 
+// Attempt to get as much information from the appChannel where possible.
+func getApplicationSourceChannel(a description.Application, url *charm.URL) (corecharm.Source, *Channel) {
+	var source corecharm.Source
 	switch url.Schema {
 	case "cs":
-		origin.Source = corecharm.CharmStore.String()
+		source = corecharm.CharmStore
 	case "local":
-		origin.Source = corecharm.Local.String()
+		source = corecharm.Local
 	default:
-		origin.Source = corecharm.CharmHub.String()
+		source = corecharm.CharmHub
+	}
+
+	c := a.Channel()
+	if c == "" {
+		return source, nil
 	}
-	return origin, nil
+
+	if source == corecharm.CharmStore || source == corecharm.Local {
+		return source, &Channel{Risk: a.Channel()}
+	}
+
+	norm, err := corecharm.ParseChannelNormalize(a.Channel())
+	if err != nil {
+		return source, nil
+	}
+
+	return source, &Channel{
+		Track:  norm.Track,
+		Risk:   string(norm.Risk),
+		Branch: norm.Branch,
+	}
+}
+
+// Attempt to locate the architecture, if it's found in the URL then use
+// that, otherwise fallback to the arch constraint.
+func (i *importer) getApplicationPlatform(a description.Application, url *charm.URL, units []description.Unit) *Platform {
+	var platform *Platform
+	if url != nil && url.Architecture != "" {
+		platform = &Platform{
+			Architecture: url.Architecture,
+			Series:       url.Series,
```

Comment:
This only works for 2.9 to 2.9 migrations, migrations from 2.8 may not have urls with series.  Fallback of a.Series?